api/src/opentrons/protocol_engine/state/geometry.py

```
@@ -977,6 +978,31 @@ def get_total_nominal_gripper_offset_for_move_type(
                     ).dropOffset
                 )
 
+    def validate_gripper_labware_tip_collision(
+        self,
+        gripper_homed_position_z: float,
+        tip: TipGeometry,
+        pipette_id: str,
+        labware_id: str,
+        current_location: OnDeckLabwareLocation,
+    ) -> bool:
+        """Check for potential collision of tips against labware to be lifted."""
+        pipette_home_z = 248.0
+        # TODO(mm, 2024-01-22): Remove the 1 and 8 channel special case once we are doing X axis validation
+        if self._pipettes.get_channels(pipette_id) in [1, 8]:
+            return True
+
+        labware_top_z_when_gripped = gripper_homed_position_z + (
+            self.get_labware_highest_z(labware_id=labware_id)
+            - self.get_labware_grip_point(
+                labware_id=labware_id, location=current_location
+            ).z
+        )
+        # TODO(mm, 2024-01-18): Utilizing the nozzle map and labware X coordinates verify if collisions will occur on the X axis (analysis will use hard coded data to measure from the gripper critical point to the pipette mount)
+        if (pipette_home_z - tip.length) < labware_top_z_when_gripped:
+            return False
+        return True
```

Instead of returning True/ False, you could have this method raise `LabwareMovementNotAllowedError` when a collision is expected. That way, you won't have to do an if/else check in `labware_movement`

---

@CaseyBatten and I talked about this and I suggested returning a bool, because our unit testing style doesn't give us a strong way of testing that exceptions propagate correctly.